feat: introducing #[internal]

benesjan opened the pull request:
Fixes F-22
Fixes #9180

Introduces `#[internal(...)]` macro that injects ContractSelf just like `#[external(...)]` macro but instead of constructing relevant context from inputs it directly accepts the context as first arg to the function (this is also injected into params by the macro). Then when a call to internal is formed via `self.internal.my_internal_func(...)` the call directly passes the relevant context.

The same approach is used as what was introduced for`#[external(...)]` in a PR down the stack in that the `#[aztec]` macro makes the original functions uncallable and generates new function with `__aztec_nr_internals__` prefix and injects it into the contract. This gives devs a nice error if they attempt to call the function directly instead of via the intended API.

benesjan commented:
* **#18417** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18417?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18413** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18413?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18396** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18396?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18398** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18398?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18289** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18289?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18274** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18274?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#18265** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18265?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17963** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17963?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#18126** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18126?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#17792** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/17792?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented on boxes/boxes/vanilla/contracts/src/main.nr:
```diff
@@ -36,7 +36,7 @@ pub contract PrivateVoting {
     #[external("private")]
     fn cast_vote(candidate: Field) {
-        let msg_sender_npk_m_hash = get_public_keys(self.context.msg_sender().unwrap()).npk_m.hash();
+        let msg_sender_npk_m_hash = get_public_keys(self.msg_sender().unwrap()).npk_m.hash();
```
Unrelated. Did a search for this when AI didn't use the new API and stumbled upon this as well.

benesjan commented on noir-projects/aztec-nr/aztec/src/macros/functions/self_call_registry.nr:
```diff
@@ -1,3 +1,6 @@
+// TODO(F-131): Drop all this and instead just iterate over the functions directly in the `#[aztec]` macro. This
+// approach was already implemented for `#[internal(...)]` functions.
```
Unrelated but realized that the registry approach used here is unnecessary and convoluted - we can just directly iterate over the functions and generate the calls accordingly. I already used this approach for generating the calls to internal. Will tackle this in a followup PR.

benesjan commented on noir-projects/noir-contracts/contracts/app/token_contract/src/main.nr:
```diff
@@ -2,7 +2,7 @@
 mod test;
 
-use dep::aztec::macros::aztec;
+use aztec::macros::aztec;
```
Cutified the imports here a bit

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/calls_generation/internal_functions.nr:
```diff
@@ -0,0 +1,81 @@
+    quote {
+        pub fn $original_function_name(self, $original_params) -> $original_return_type {
+            $fn_name(self.context, $params_at_callsite)
+        }
```
```suggestion
        pub fn $original_function_name(self: CallInternal, $original_params) -> $original_return_type {
            $fn_name(self.context, $params_at_callsite)
        }
```

Thsi might make the macro easier to read

benesjan replied:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/17963/commits/8b35b940b25c0fd82ae3806060767f9165198448

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/calls_generation/internal_functions.nr:
```diff
@@ -0,0 +1,81 @@
+
+use crate::macros::internals_functions_generation::internal_functions_registry;
+
+comptime fn generate_private_internal_function_call(f: FunctionDefinition) -> Quoted {
```
```suggestion
/// Generates a method for the `CallInternal` struct that makes a call to the `#[internal]` function `f`.
comptime fn generate_private_internal_function_call(f: FunctionDefinition) -> Quoted {
```

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/mod.nr:
```diff
@@ -159,19 +159,51 @@ pub comptime fn external(f: FunctionDefinition, f_type: CtString) {
+    let function_name = f.name();
+    if is_fn_external(f) {
+        panic(
+            f"The #[internal] attribute cannot be applied to external functions - {function_name} is marked as both #[external(\"{f_type}\")] and #[internal]",
```
```suggestion
            f"The #[internal] attribute cannot be applied to external functions - {function_name} is marked as both #[external] and #[internal(\"{f_type}\")]",
```

nventuro commented on noir-projects/aztec-nr/aztec/src/macros/functions/utils.nr:
```diff
@@ -528,9 +530,12 @@ pub(crate) comptime fn create_authorize_once_check(
+            & !is_fn_test(f) {
             panic(
-                f"Function {name} must be marked as either #[external(...)], #[contract_library_method], or #[test]",
+                f"Function {name} must be marked as either #[external(...)], #[contract_library_method], #[internal(...)], or #[test]",
```
```suggestion
// We  don't suggest that #[contract_library_method] is allowed because we don't want to introduce another
// concept
panic(
                f"Function {name} must be marked as either #[external(...)], #[internal(...)], or #[test]",
```

Also drop usage from our examples.

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/17963/commits/35153b148cc4d1372b1b224da5a592e1d9ad42db

benesjan committed:
feat: introducing #[internal]

Fixes F-22
Fixes #9180

Introduces `#[internal(...)]` macro that injects ContractSelf just like `#[external(...)]` macro but instead of constructing relevant context from inputs it directly accepts the context as first arg to the function (this is also injected into params by the macro). Then when a call to internal is formed via `self.internal.my_internal_func(...)` the call directly passes the relevant context.

The same approach is used as what was introduced for`#[external(...)]` in a PR down the stack in that the `#[aztec]` macro makes the original functions uncallable and generates new function with `__aztec_nr_internals__` prefix and injects it into the contract. This gives devs a nice error if they attempt to call the function directly instead of via the intended API.
